analysis/semsim/unifiedmodel.py

```diff
@@ -82,3 +88,83 @@ def _tokenize(self, example):
                 features[f"{example_key}_{feat_key}"] = features.pop(feat_key)
             features_full.update(features)
         return features_full
+
+
+class RRUM(pl.LightningModule):
+    def __init__(self, with_transcript, optimizer_config, freeze_policy=None, pos_weight=None):
+        super().__init__()
+        self.optimizer_config = optimizer_config
+        self._text_types = ['title', 'description'] + \
+            (['transcript'] if with_transcript else [])
+        self.cross_encoders = nn.ModuleDict({})
+        for t in self._text_types:
+            self.cross_encoders[t] = AutoModelForSequenceClassification.from_pretrained(
+                'cross-encoder/stsb-roberta-base').to("cuda:0")
+        if freeze_policy is not None:
+            for xe in self.cross_encoders.values():
+                for name, param in xe.named_parameters():
+                    if freeze_policy(name):
+                        param.requires_grad = False
+
+        self.lin1 = nn.Linear(len(self._text_types) + 1, 1)
+        self.ac_metric = torchmetrics.Accuracy()
+        self.pr_metric = torchmetrics.Precision()
+        self.re_metric = torchmetrics.Recall()
+        self.auc_metric = torchmetrics.AUROC()
+
+        if pos_weight:
+            self.loss = nn.BCEWithLogitsLoss(
+                pos_weight=torch.Tensor([pos_weight]))
+        else:
+            self.loss = nn.BCEWithLogitsLoss()
+
+    def forward(self, x):
+        # for some reason the prediction call wraps the inputs in a length-1 list
+        if isinstance(x, list):
+            if (len(x) != 1):
+                print("BAD!!!")
+            x = x[0]
+        cross_logits = {}
+        for f in self._text_types:
+            cross_logits[f] = self.cross_encoders[f](
+                input_ids=x[f"{f}_input_ids"], attention_mask=x[f"{f}_attention_mask"]).logits
+        scalar = x["channel_sim"]
+        x = torch.cat((
+            *cross_logits.values(),
+            scalar[:, None]),
+            1
+        )
+        del cross_logits, scalar
+
+        x = self.lin1(x)
+        return x
+
+    def configure_optimizers(self):
+        return self.optimizer_config(self)
```

Comment:
If we change the optimizer config as in the init comment, then this could be changed to
`return self.optimizer_class(self.parameters(), lr=self.lr)`